### ISSUE TITLE
Add LR Scheduler to full finetune recipes

### ISSUE THREAD

**Issue by joecummings** [pytorch/torchtune#1308]
Following [how we do it in LoRA recipes](https://github.com/pytorch/torchtune/blob/afd23fd0b2f9051958affae20890396e2594756f/recipes/lora_finetune_distributed.py#L475), we should add the ability to use a LR scheduler to the full finetune recipes

Here's everything that would need to be changed:
- [ ] Add appropriate functions to recipes/full_finetune_distributed.py (follow recipes/lora_finetune_distribute.py for example)
- [x] Add appropriate functions to recipes/full_finetune_single_device.py (follow recipes/lora_finetune_single_device.py for example)
- Run an example run with Llama3.1 8B and log to Weights & Biases. Post screenshot confirming this works in your PR

I don't think we need to update all our configs yet, just need to make sure the functionality exists.

**Comment by felipemello1** [comment]
@joecummings , I was thinking that this could be a function in utils that takes an optimizer as input, and just does the step for, that way it can be shared for any recipe. Maybe it need a flag for "is_distributed".

This complicates the task, but maybe we could follow a similar pattern for model checkpointing and early stopping, which are also frequently asked features.

**Comment by joecummings** [comment]
We CaN aLWayS dO iT IN a FollOw Up REfaCtOR

**Comment by joecummings** [comment]
Kidding, I know that will never happen. I like the idea.

**Comment by user074** [comment]
I guess it kinda work for the full_finetune_single_device.py. I will update after I finish the training to see how it goes
<img width="1245" alt="Screenshot 2024-08-13 at 12 39 08 AM" src="https://github.com/user-attachments/assets/59e8f7bf-3631-40b1-92da-bf9983c5bdd7">

**Comment by SalmanMohammadi** [comment]
Can we close this?

**Comment by ebsmothers** [comment]
@SalmanMohammadi technically I think we still need to add to the distributed recipe to fully close it out

**Comment by user074** [comment]
I can have a check on the distributed case but probably after mid November

**Comment by felipemello1** [comment]
fyi, reference implementation from a fork: https://github.com/homebrewltd/torchtune/blob/1ff4fc217e578a1656dc60c343dcee98e7ac013e/recipes/full_finetune_fsdp2.py#L285
